### PR TITLE
fix(rest): show all linked releases with project relationship information

### PR DESCRIPTION
**Pull request by maierthomas**
- All releases for a project are now displayed
- Added `linkRelation` to the self link of linked releases

```
 "containedReleases": [
      {
        "name": "Release ABC",
        "version": "1.2.3",
        "clearingState": "APPROVED",
        "_links": {
          "self": {
            "href": "https://localhost:8443/resource/api/releases/1234567.....",
            "linkRelation": {
              "releaseRelation": "CONTAINED",
              "mainlineState": "OPEN"
            }
          }
        }
      }
    ]
```

closes sw360/sw360portal#718
closes sw360/sw360portal#722

### CODE REVIEW

**Comment by maxhbr** [comment]
I would expect `"linkRelation"` on the same level as `"name"`. In my understanding the filed `"_links"` has a clear structure and is mostly parsed automatically. Adding data there does not look like the best idea to me.

maybe ping @toedter

**Comment by adrian-evo** [comment]
Works for me as initially expected.

**Comment by maierthomas** [comment]
Ok maybe we should leave the self link untouched.

For your first suggestion we need to wrap or extend the release somehow because currently we can not store the linked release information.
I dont really like the second suggestion because the linked information should be in the same element with the release.

**Comment by mcjaeger** [comment]
offtopic: Boeing 720 was a nice plane by the way: https://en.wikipedia.org/wiki/Boeing_720 :-)

**Comment by mcjaeger** [comment]
in discussion we pointed out preference for the variant below:

* similar to sw360 data model
* simple

  ```
   "containedReleases": [
        {
          "releaseRelation": "CONTAINED",
          "mainlineState": "OPEN",
          "release": {
            "name": "Release ABC",
            "version": "1.2.3",
            "clearingState": "APPROVED",
            "_links": {
              "self": {
                "href": "https://localhost:8443/resource/api/releases/1234567.....",
              }
            }
          }
        }
      ]
  ```

**Comment by maxhbr** [comment]
```
 "containedReleases": [
      {
        "releaseRelation": "CONTAINED",
        "mainlineState": "OPEN",
        "release": {
            "self": {
              "href": "https://localhost:8443/resource/api/releases/1234567.....",
          }
      }
    ]
"_embedded": [ 
{

          "name": "Release ABC",
          "version": "1.2.3",
          "clearingState": "APPROVED",
          "_links": {
            "self": {
              "href": "https://localhost:8443/resource/api/releases/1234567.....",
            }
          }
        }
]
```

**Comment by maierthomas** [comment]
Current implementation: 
We can use the linkedProject and releaseIdToUsage maps to show the relevant link relation information.
All the releases are embedded.
```

{
  "type": "project",
  "name": "abc",
  "description": "",
  "version": "V1",
  "externalIds": {
    "xxx-mainline-id-project": "123456"
  },
  "createdOn": "2016-10-10",
  "businessUnit": "EE",
  "linkedProjects": {
    "065f3aa45c6683297fd1bb3929813c94": "CONTAINED",
    "065f3aa45c6683297fd1bb39297a0c24": "CONTAINED"
  },
  "linkedReleases": {
    "065f3aa45c6683297fd1bb3929026bd1": {
      "releaseRelation": "CONTAINED",
      "mainlineState": "OPEN"
    },
    "065f3aa45c6683297fd1bb3929087f4f": {
      "releaseRelation": "CONTAINED",
      "mainlineState": "OPEN"
    },
    "065f3aa45c6683297fd1bb39291add64": {
      "releaseRelation": "CONTAINED",
      "mainlineState": "OPEN"
    }
  },
  "deliveryStart": "",
  "projectType": "PRODUCT",
  "_links": {
    "self": {
      "href": "https://localhost:8443/resource/api/projects/065f3aa45c6683297fd1bb3929813219"
    }
  },
  "_embedded": {
    "createdBy": {
      "email": "admin@sw360.org",
      "_links": {
        "self": {
          "href": "https://localhost:8443/resource/api/users/YWRtaW5Ac3czNjAub3Jn"
        }
      }
    },
    "sw360:moderators": [
      {
        "email": "name.a@sw360.org",
        "_links": {
          "self": {
            "href": "https://localhost:8443/resource/api/users/dGhvbWFzLncuc2NobWlkdEBzaWVtZW5zLmNvbQ=="
          }
        }
      },
      {
        "email": "name.b@sw360.org",
        "_links": {
          "self": {
            "href": "https://localhost:8443/resource/api/users/YXVzdGluLmxhbUBzaWVtZW5zLmNvbQ=="
          }
        }
      },
      {
        "email": "name.c@sw360.org",
        "_links": {
          "self": {
            "href": "https://localhost:8443/resource/api/users/amFuLmVoQHNpZW1lbnMuY29t"
          }
        }
      }
    ],
    "sw360:releases": [
      {
        "name": "Release Name 1",
        "version": "1.1.0",
        "clearingState": "APPROVED",
        "_links": {
          "self": {
            "href": "https://localhost:8443/resource/api/releases/065f3aa45c6683297fd1bb3929026bd1"
          }
        }
      },
      {
        "name": "Release Name 2",
        "version": "1.2.0",
        "clearingState": "APPROVED",
        "_links": {
          "self": {
            "href": "https://localhost:8443/resource/api/releases/065f3aa45c6683297fd1bb3929087f4f"
          }
        }
      },
      {
        "name": "Release Name 3",
        "version": "2.3.1",
        "clearingState": "APPROVED",
        "_links": {
          "self": {
            "href": "https://localhost:8443/resource/api/releases/065f3aa45c6683297fd1bb39291add64"
          }
        }
      }
    ]
  }
}
```

**Comment by mcjaeger** [comment]
```
  "createdOn": "2016-10-10",
  "businessUnit": "EE",
  "linkedProjects": {
    "065f3aa45c6683297fd1bb3929813c11":  {
      "releaseRelation": "CONTAINED",
      "href": "https://localhost:8443/resource/api/projects/065f3aa45c6683297fd1bb3929813c11"
    },
    "065f3aa45c6683297fd1bb39297a0c24": "CONTAINED"
  },
  "linkedReleases": {
    "065f3aa45c6683297fd1bb3929026bd1": {
      "releaseRelation": "CONTAINED",
      "mainlineState": "OPEN"
    },
    "065f3aa45c6683297fd1bb3929087f4f": {
      "releaseRelation": "CONTAINED",
      "mainlineState": "OPEN"
    },
    "065f3aa45c6683297fd1bb39291add64": {
      "releaseRelation": "CONTAINED",
      "mainlineState": "OPEN"
      "href": "https://localhost:8443/resource/api/releases/065f3aa45c6683297fd1bb39291add64"
    }
  },
  "deliveryStart": "",
  "projectType": "PRODUCT",
  "_links": {
    "self": {
      "href": "https://localhost:8443/resource/api/projects/065f3aa45c6683297fd1bb3929813219"
    }
  },
```

**Comment by maierthomas** [comment]
Created serializer for linked projects and releases

```
{
  "type": "project",
  "name": "Test Project",
  "description": "",
  "version": "V04.75",
  "externalIds": {
    "a-mainline-id-project": "1234"
  },
  "createdOn": "2016-10-10",
  "businessUnit": "GD",
  "deliveryStart": "",
  "projectType": "PRODUCT",
  "linkedProjects": [
    {
      "project": "https://localhost:8443/resource/api/projects/065f3aa45c6683297fd1bb39297b87bd",
      "relation": "CONTAINED"
    },
    {
      "project": "https://localhost:8443/resource/api/projects/065f3aa45c6683297fd1bb39297bafee",
      "relation": "CONTAINED"
    }
  ],
  "linkedReleases": [
    {
      "release": "https://localhost:8443/resource/api/releases/065f3aa45c6683297fd1bb392913b1fc",
      "mainlineState": "OPEN",
      "relation": "CONTAINED"
    },
    {
      "release": "https://localhost:8443/resource/api/releases/065f3aa45c6683297fd1bb392917ac71",
      "mainlineState": "OPEN",
      "relation": "CONTAINED"
    },
    {
      "release": "https://localhost:8443/resource/api/releases/065f3aa45c6683297fd1bb39291794ff",
      "mainlineState": "OPEN",
      "relation": "CONTAINED"
    }
  ],
  "_links": {
    "self": {
      "href": "https://localhost:8443/resource/api/projects/065f3aa45c6683297fd1bb39297b88cb"
    }
  },
  "_embedded": {
    "createdBy": {
      "email": "admin@sw360.org",
      "_links": {
        "self": {
          "href": "https://localhost:8443/resource/api/users/YWRtaW5Ac3czNjAub3Jn"
        }
      }
    },
    "sw360:moderators": [
      {
        "email": "a@sw360.org",
        "_links": {
          "self": {
            "href": "https://localhost:8443/resource/api/users/dGhvbWFzLncuc2NobWlkdEBzaWVtZW5zLmNvbQ=="
          }
        }
      },
      {
        "email": "b@sw360.org",
        "_links": {
          "self": {
            "href": "https://localhost:8443/resource/api/users/amFuLmVoQHNpZW1lbnMuY29t"
          }
        }
      }
    ],
    "sw360:releases": [
      {
        "name": "test release",
        "version": "5.1.25",
        "_links": {
          "self": {
            "href": "https://localhost:8443/resource/api/releases/065f3aa45c6683297fd1bb392913b1fc"
          }
        }
      },
      {
        "name": "test release 2",
        "version": "12",
        "_links": {
          "self": {
            "href": "https://localhost:8443/resource/api/releases/065f3aa45c6683297fd1bb392917ac71"
          }
        }
      },
      {
        "name": "test release 3",
        "version": "4",
        "_links": {
          "self": {
            "href": "https://localhost:8443/resource/api/releases/065f3aa45c6683297fd1bb39291434ae"
          }
        }
      },
      {
        "name": "test release 64",
        "version": "3.7.6",
        "_links": {
          "self": {
            "href": "https://localhost:8443/resource/api/releases/065f3aa45c6683297fd1bb39291794ff"
          }
        }
      }
    ]
  }
}
```

**Comment by mcjaeger** [comment]
@adrian-evo would it be OK for you to test please?

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/RestControllerHelper.java [hunk 1]
```diff
@@ -243,7 +242,30 @@ private void addEmbeddedAttachments(
...
+    private void addEmbeddedProject(HalResource halResource, Project project) {
+        Project minimalFilledProject = new Project(project.getName());
+        minimalFilledProject.setType(null);
+        minimalFilledProject.setProjectType(null);
+        HalResource<Project> halProject = new HalResource<>(minimalFilledProject);
+
+        try {
+            Link projectLink = linkTo(ProjectController.class).slash("api" + ProjectController.PROJECTS_URL + "/" + project.getId()).withSelfRel();
+            halProject.add(projectLink);
+        } catch (Exception e) {
```
Comment (alexbrdn) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/RestControllerHelper.java]: why the generic `catch`? can't it be removed?

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/ProjectController.java [hunk 2]
```diff
@@ -257,18 +260,15 @@ public RepositoryLinksResource process(RepositoryLinksResource resource) {
...
+        Map<String, ProjectReleaseRelationship> releaseIdToUsage = sw360Project.getReleaseIdToUsage();
+        if (!Objects.isNull(releaseIdToUsage)) {
+            restControllerHelper.addEmbeddedReleases(halProject, releaseIdToUsage.keySet(), releaseService, sw360User);
+        }
+        Map<String, ProjectRelationship> linkedProjects = sw360Project.getLinkedProjects();
+        if (!Objects.isNull(linkedProjects)) {
+            restControllerHelper.addEmbeddedProject(halProject, linkedProjects.keySet(), projectService, sw360User);
         }
-        if (sw360Project.getModerators() != null) {
+        if (!Objects.isNull(sw360Project.getModerators())) {
```
Comment (alexbrdn) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/ProjectController.java]: Why did you change a perfectly readable code to this contortion? (Here and above as well)

Quote from the javadoc:

> This method exists to be used as a Predicate, filter(Objects::isNull)

And then there's the method `Objects.nonNull()`... 

I really don't see the point. You're not making the code any more readable and you're misusing the helper method.

Comment (maierthomas) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/ProjectController.java]: stupid habit... corrected to !=null

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/RestControllerHelper.java (hunk 1 above)
Comment (maierthomas) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/RestControllerHelper.java]: its not necessary here... copied code. corrected all occurrences

**Comment by adrian-evo** [comment]
Works as expected.
